course/api/lti_views.py

```diff
@@ -0,0 +1,127 @@
+        # If need arises, they may be added.
+
+
+class CourseLineItemsViewSet(viewsets.GenericViewSet, OAuth2ScopeChecker):
+    """
+    ViewSet for reading LTI line items associated with a given course.
+    Also implements 'scores' action for posting grades related to a line item.
+    """
+    parent_lookup_map = {'course_id': 'id'}
+    lookup_field = 'exercise_id'
+    lookup_url_kwarg = 'exercise_id'
+    lookup_value_regex = REGEX_INT
+    serializer_class = LtiLineItemSerializer
+    queryset = LTI1p3Exercise.objects.all()
+    permission_classes = [OAuth2TokenPermission]
+
+    def check_token_scope(self, scopes: List[str]) -> bool:
+        if self.action == 'scores':
+            if 'https://purl.imsglobal.org/spec/lti-ags/scope/score' in scopes:
+                return True
+        else:
+            if 'https://purl.imsglobal.org/spec/lti-ags/scope/lineitem.readonly' in scopes:
+                return True
+        return False
+
+    def list(self, request: Request, *args, **kwargs):
+        queryset = self.queryset.filter(course_module__course_instance__pk=kwargs['course_id'])
+        result = []
+        for exercise in queryset:
+            lineitem = LineItem(request, exercise, kwargs['course_id'])
+            serializer = LtiLineItemSerializer(lineitem)
+            result.append(serializer.data)
+
+        return Response(result)
+
+    def retrieve(self, request: Request, *args, **kwargs):
+        try:
+            exercise = LTI1p3Exercise.objects.get(pk=kwargs['exercise_id'])
+            lineitem = LineItem(request, exercise, kwargs['course_id'])
+            serializer = LtiLineItemSerializer(lineitem)
+        except LTI1p3Exercise.DoesNotExist as e:
+            logger.error("LTI 1.3/lineitem: Exercise (%d) not found", kwargs['exercise_id'])
+            raise Http404("LTI exercise not found") from e
+        return Response(serializer.data)
+
+    @action(methods=['post'], detail=True, url_path='scores')
+    def scores(self, request: Request, *args, **kwargs):
+        jsondata = json.loads(request.body)
+        serializer = LtiScoresSerializer(data=jsondata)
+        serializer.is_valid()
+        data = serializer.validated_data
+
+        try:
+            exercise = LTI1p3Exercise.objects.get(pk=kwargs['exercise_id'])
+        except LTI1p3Exercise.DoesNotExist as e:
+            logger.error("LTI 1.3/scores: Exercise (%d) not found", kwargs['exercise_id'])
+            raise Http404("LTI exercise not found") from e
+
+        try:
+            user = User.objects.get(pk=int(data.get('userId')))
+        except UserProfile.DoesNotExist:
+            logger.error("LTI 1.3/scores: User (%s) does not exist", data.get('userId'))
+            return Response({'UserProfile does not exist'}, status=400)
+
+        logger.info("Received LTI 1.3 'scores' request for exercise %s, user %s", str(exercise), str(user))
+        sub = Submission.objects.filter(exercise=exercise, submitters=user.userprofile)
```

Comment:
Does this match submissions that have multiple submitters and `user.userprofile` is one of them?

Reply:
What would be the most convenient way of handling the multiple submitters case? On the other hand, with LTI 1.3 the submissions always have just single submitter, and I don't think LTI 1.3 protocol even supports multi-user use.

Reply:
Based on the django tests https://github.com/django/django/blob/531f557f9238b8f2e5032e569cf36f0c05bb4043/tests/many_to_many/tests.py#L207-L210, the code indeed matches all submission that have `user.userprofile` in `submitters`.

Reply:
Since LTI doesn't support groups, let's just assume that we don't need do anything special about them. Leave the code as is.